In [1]:
# Q1: Demo matrices - 1 point

# Assuming a Stride of 1, the Output matrix will be:
# 4 3 4
# 2 4 3
# 2 3 4

Calculate the output:

Image:

1 1 1 0 0

0 1 1 1 0

0 0 1 1 1

0 0 1 1 0

0 1 1 0 0


Filter:

1 0 1

0 1 0

1 0 1

In [2]:
#Q2: What are the advantages of a CNN over a fully connected DNN for image classification? - 1 point

- They are able to capture and learn relevant features from an image at different levels while a DNN cannot (feature learning)
- CNN networks can be less complex than DNN through the use of filters and weight sharing: each neuron does not need to be fully connected
- This lowering of neuron connections allow the CNN to be trained much faster than a DNN

In [3]:
#Q3: If your GPU runs out of memory while training a CNN, what are five things you could try to solve the problem? - 1 point

- 1) Try to rescale the images down in size so that they don't take up that much memory
- 2) Lower the batch size for training so that fewer images are trained during each batch
- 3) Decrease the number of layers within the CNN itself so that fewer parameters are being used
- 4) Use Pooling layers within the network to downsample the image to a smaller size
- 5) Buy a better GPU :) - Kidding - Adjust the filter output size so that it's smaller than the current one

In [4]:
#Q4: Use the dataset: https://www.tensorflow.org/datasets/catalog/citrus_leaves - 1 point
# Using the Cifar10 dataset since citrus leaves has been not working
import tensorflow as tf
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.cifar10.load_data()

# Normalize pixel values to be between 0 and 1
train_images, test_images = train_images / 255.0, test_images / 255.0

#Q5: Split it into a training set, and a test set. - 1 point
# Above dataset is already loaded as train and test data with 50,000 images for train
# and 10,000 for test

170500096/170498071 [==============================] - 13s 0us/step


In [6]:
from keras import models
from keras import layers

In [36]:
#Q6: Build a CNN model - 3 points


model = models.Sequential()
# Adding a 2D convolution layer and specifying the input shape of the matrix
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
# Max pool layer for downsizing and specifying size of output
model.add(layers.MaxPooling2D((2, 2)))
# Another layer - specifying neuron size and matrix output
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
# Flattening 3D image down to 1D for classification
model.add(layers.Flatten())
#model.add(layers.Dense(64, activation='relu'))
# Adding dense layer as final output for classes
model.add(layers.Dense(10))

model.summary()

# Adam optimizer for fast training and SparseCatCross for image loss calculation
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
# Credit to: https://www.tensorflow.org/tutorials/images/cnn

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 30, 30, 32)        896       
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 15, 15, 32)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 13, 13, 64)        18496     
_________________________________________________________________
flatten (Flatten)            (None, 10816)             0         
_________________________________________________________________
dense (Dense)                (None, 10)                108170    
Total params: 127,562
Trainable params: 127,562
Non-trainable params: 0
_________________________________________________________________


In [37]:
model.fit(train_images, train_labels, epochs=5, validation_split=.1)

Epoch 1/5
1407/1407 [==============================] - 19s 13ms/step - loss: 2.2286 - accuracy: 0.3204 - val_loss: 1.5778 - val_accuracy: 0.4322
Epoch 2/5
1407/1407 [==============================] - 21s 15ms/step - loss: 1.4369 - accuracy: 0.4932 - val_loss: 1.4633 - val_accuracy: 0.4816
Epoch 3/5
1407/1407 [==============================] - 24s 17ms/step - loss: 1.2720 - accuracy: 0.5600 - val_loss: 1.3106 - val_accuracy: 0.5440
Epoch 4/5
1407/1407 [==============================] - 26s 19ms/step - loss: 1.1061 - accuracy: 0.6180 - val_loss: 1.2950 - val_accuracy: 0.5608
Epoch 5/5
1407/1407 [==============================] - 28s 20ms/step - loss: 0.9915 - accuracy: 0.6602 - val_loss: 1.3333 - val_accuracy: 0.5762


In [38]:
test_loss, test_acc = model.evaluate(test_images,  test_labels)
print("Model accuracy: ",test_acc)

313/313 [==============================] - 1s 3ms/step - loss: 1.3863 - accuracy: 0.5672
Model accuracy:  0.5672000050544739


In [7]:
#Q7: Fine-tune it and learn its accuracy. Bring the minimum accuracy of 85%. Jot down any challenges/issues you face during this process. - 3 points

model = models.Sequential()
# Adding a 2D convolution layer and specifying the input shape of the matrix
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
# Max pool layer for downsizing and specifying size of output
model.add(layers.MaxPooling2D((2, 2)))
# Another layer - specifying neuron size and matrix output
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.BatchNormalization())
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.BatchNormalization())
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.BatchNormalization())
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.BatchNormalization())
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Flattening 3D image down to 1D for classification
model.add(layers.Flatten())
# Adding hidden dense layer before output layer
model.add(layers.Dense(64, activation='relu'))
# Adding dense layer as final output for classes
model.add(layers.Dense(10))

model.summary()
# Adam optimizer for fast training and SparseCatCross for image loss calculation
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 30, 30, 32)        896       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 15, 15, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 13, 13, 64)        18496     
_________________________________________________________________
batch_normalization (BatchNo (None, 13, 13, 64)        256       
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        36928     
_________________________________________________________________
batch_normalization_1 (Batch (None, 11, 11, 64)        256       
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 9, 9, 64)          3

In [8]:
model.fit(train_images, train_labels, epochs=15, validation_split=.1)

Epoch 1/15
1407/1407 [==============================] - 63s 45ms/step - loss: 1.5499 - accuracy: 0.4404 - val_loss: 1.6341 - val_accuracy: 0.4422
Epoch 2/15
1407/1407 [==============================] - 60s 43ms/step - loss: 1.1887 - accuracy: 0.5786 - val_loss: 1.1929 - val_accuracy: 0.5632
Epoch 3/15
1407/1407 [==============================] - 66s 47ms/step - loss: 1.0209 - accuracy: 0.6420 - val_loss: 1.0038 - val_accuracy: 0.6478
Epoch 4/15
1407/1407 [==============================] - 67s 48ms/step - loss: 0.9016 - accuracy: 0.6818 - val_loss: 0.9853 - val_accuracy: 0.6578
Epoch 5/15
1407/1407 [==============================] - 62s 44ms/step - loss: 0.8074 - accuracy: 0.7177 - val_loss: 0.9998 - val_accuracy: 0.6572
Epoch 6/15
1407/1407 [==============================] - 62s 44ms/step - loss: 0.7258 - accuracy: 0.7434 - val_loss: 1.0821 - val_accuracy: 0.6240
Epoch 7/15
1407/1407 [==============================] - 66s 47ms/step - loss: 0.6612 - accuracy: 0.7684 - val_loss: 1.1956 -

In [9]:
test_loss, test_acc = model.evaluate(test_images,  test_labels)
print("Model accuracy: ",test_acc)

313/313 [==============================] - 2s 8ms/step - loss: 1.2090 - accuracy: 0.6863
Model accuracy:  0.6862999796867371


## Challenges/Process to 85%
- First iteration was base model at Q6. Accuracy was low for training and even lower for testing (56%). Attempt to add more convolution layers and train for more epochs
- Second attempt was slightly better by adding two more convolution layers to model. Test accuracy at 60%. Will try adding dense layer after the flatten
- Adding extra dense layer actually made accuracy drop to 58%. Going to normalize the data to between 0 and 1 and add more convolution layers to model. Also running for 15 epochs now.
- Running at 15 epochs did help in raising accuracy to 67% along with more covolution layers and the normalizing. Now adding BatchNormalization layers to model so that the feature space be kept tight.
- Final run was able to raise accuracy to 68%. I could not get it to 85% but if I had more time I might be able to. I was able to get training accuracy to 88% but clearly there is overfitting going on since validation accuracy is much lower. If I had more time, I would experiment with adding some dropout layers into the model.

Useful reading: 
*   https://www.tensorflow.org/tutorials/generative/style_transfer

*   https://towardsdatascience.com/build-your-own-convolution-neural-network-in-5-mins-4217c2cf964f